### PR TITLE
Dynamicmarket

### PR DESCRIPTION
**Pull request by dbenson24**
This PR is pretty close to being finished. 

Biggest missing piece is integration with BattleEye and SpyGlass

### CODE REVIEW

**Comment by CarlosHingel** [comment]
Finally! Oh yeah! Thank you guys :P

@dbenson24 

I`ve been using your dynamic market on my server for some time now and its amazing! But there is a bug that my players are informing me..... The prices are dropping to fast, and i mean that when a player is selling his shit the prices drop the moment he sells, in other words, he will sell the first part of his farm for the price on the market, but when he tries to sell the second and\or third part, the prices are already dropping and it continues to drop, even if you set in the config a delay of like 8 minutes for the prices to hit the market.

**Comment by Jawshy** [comment]
you-got-anymore-of-them-commits-compressed

Don't forget to squad and merge!

**Comment by dbenson24** [comment]
@CarlosHingel Did you properly setup the server restart time? If it thinks the server should have restarted already all market changes immediately take affect

**Comment by setoy** [comment]
My restart is every 6 hours. I've got the same problem like @CarlosHingel 


    end_time = 21600; //Last time to sync prices (Default: 4 hours, 14400 in seconds)
    sync_rate = 900; //Seconds between syncs to database (Default: 15 minutes, in seconds)

**Comment by tkcjesse** [comment]
I'm gonna help you clean up stuff yo. We'll wait to merge till then probably. If @BoGuu agrees.

**Comment by tkcjesse** [comment]
@dbenson24,

I've submitted a PR - https://github.com/dbenson24/Framework/pull/3

I work the next three days or so. I'll slowly add to it as I can. However, we do need localization done for a lot of things in there just as a heads up.

**Comment by CarlosHingel** [comment]
This feature is a game play change one! 

On my server is so nice seeing gangs divided themselves into groups to farm different types of resources and it also make them use more the houses of the game to store their farm so when the price is high they will sell it and it also make the "break" into a house more meaningful because players will not only store weapons there but resources also.

**Comment by Jawshy** [comment]
> Biggest missing piece is integration with BattleEye and SpyGlass

I can update the BattlEye filters and SpyGlass variables when the rest of the changes are ready to be merged.

**Comment by coresync2k** [comment]
Will this Eco System, save the last Prices from the Server into the Database?

**Comment by tkcjesse** [comment]
@dbenson24 Okay, so an issue I just thought of while doing some stuff with this. If you set ```dynamic_market = false;```.. it still requires you to have your prices set via the economy table in the DB yes?

**Comment by dbenson24** [comment]
@coresync2k Prices are saved to the database if you want them to be, I have a config option for price changes to be dynamic only over server restarts.

@tkcjesse Yea, prices are set via the database regardless of whether or the market system is used. It wouldn't be too hard to change it so that when the market is disabled prices are loaded from the config file. Or we could just enforce that prices are set in the database

**Comment by coresync2k** [comment]
@dbenson24 that would be nice, with server prices over restart.

**Comment by tkcjesse** [comment]
@dbenson24, yes that is an issue imo. The config maintains the same settings. Ideally, what I, note I.. would like to see is basically 3 different versions in one. 

1. User can disable dynamic market and be able to remove the table if they want and not have any file depend on the table even being there. They simply use the config file to handle all prices as it currently works. So basically nothing changed from now.

2. User can disable dynamic market over restarts. Meaning each server restart the market is back to normal. Yes, some people like this... 

3. User can enable market over restarts and have prices consistent.

**However...**

**Comment by dbenson24** [comment]
@tkcjesse I like that idea of keeping min/max prices in the config file. My internship ends this week so I'll take some time to implement some of these changes next week

File: Altis_Life.Altis/dialog/MasterHandler.hpp [hunk 1]
```diff
@@ -27,4 +27,6 @@
 #include "federalReserve.hpp"
 #include "compensate.hpp"
 #include "FuelStat.hpp"
-#include "newsStation.hpp"
\ No newline at end of file
+#include "AltisPhone.hpp"
```
Comment (DarkSilencerZA) [inline Altis_Life.Altis/dialog/MasterHandler.hpp]: This was removed @dbenson24  .. make sure you have the most updated files, cause it doesn't look like it :<

**Comment by DarkSilencerZA** [comment]
See here: https://github.com/AsYetUntitled/Framework/pull/47/files#diff-58e2b472a9aadc453ed9803665c03e4fR30

File: Altis_Life.Altis/dialog/MasterHandler.hpp (hunk 1 above)
Comment (dbenson24) [inline Altis_Life.Altis/dialog/MasterHandler.hpp]: Fixed. My bad, I messed up the merge

**Comment by dbenson24** [comment]
Config File now handles grouping/max/min/var prices. Added option to have non persistent dynamic market

**Comment by coresync2k** [comment]
really nice work, i will test it later on my Dev-Server

**Comment by coresync2k** [comment]
Hello, i got everytime the error -> 

`18:27:25 Error in expression < >> "factor"),
_sellPrice,
_sellPrice
]
missionNamespace setVariable [_name, _in>
18:27:25   Error position: <missionNamespace setVariable [_name, _in>
18:27:25   Error Missing ;
18:27:25 File life_server\Functions\Economy\fn_loadPrices.sqf, line 45
18:27:25 Error in expression < >> "factor"),
_sellPrice,
_sellPrice
]
missionNamespace setVariable [_name, _in>
18:27:25   Error position: <missionNamespace setVariable [_name, _in>
18:27:25   Error Missing ;
18:27:25 File life_server\Functions\Economy\fn_loadPrices.sqf, line 45`

@line 31 of this file misses a " ; " right?
but this dont fix this problem ^^

**Comment by dbenson24** [comment]
@coresync2k I missed a ; at the end of line 46. Should go after the right bracket that closes the array

**Comment by coresync2k** [comment]
@Line 23 ->     _sellPrice = _x select 1, <- needs a ; 
after fixxing this error msg from "syncprices" line 25

2016/08/07, 18:48:37 Error in expression <;
};

_goods = [];

{
if ((_x select 1) != 0) then {
_name = format["%1MarketGoo>
2016/08/07, 18:48:37   Error position: <!= 0) then {
_name = format["%1MarketGoo>
2016/08/07, 18:48:37   Error Generic error in expression
2016/08/07, 18:48:37 File life_server\Functions\Economy\fn_syncPrices.sqf, line 25

**Comment by Jawshy** [comment]
@coresync2k, you must not realise this, but every time you write "[at]line" you're [mentioning](https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-users-and-teams) another organisation. Also, [format code blocks](https://help.github.com/articles/creating-and-highlighting-code-blocks/) by placing triple backticks <code>```</code> before and after code.

**Comment by Jawshy** [comment]
The following is written to the client/local.rpt log when attempting to access a virtual item shop (market): 

```
13:45:18 Error in expression <ariable (_varname);
_itemArray pushBack _priceArray;
} forEach _shopItems;
} els>
13:45:18   Error position: <_priceArray;
} forEach _shopItems;
} els>
13:45:18   Error Undefined variable in expression: _pricearray
13:45:18 File mpmissions\__CUR_MP.Altis\core\economy\fn_retrievePrices.sqf, line 38
```

I didn't get an error in expression with fn_loadPrices.sqf. 

The error in the server.rpt was: 
```
2016/08/08, 13:33:15 Error in expression <ect 0),(_price select 2)];
};
} forEach _market;

{
_name = (_x select 0);
_pric>
2016/08/08, 13:33:15   Error position: <_market;

{
_name = (_x select 0);
_pric>
2016/08/08, 13:33:15   Error Undefined variable in expression: _market
2016/08/08, 13:33:15 File life_server\Functions\Economy\fn_syncPrices.sqf, line 31
```
----

As of https://github.com/AsYetUntitled/Framework/pull/47/commits/4896cccaeccb253ae58e9af039d101a90196fd85.

**Comment by dbenson24** [comment]
@Jawshy are you sure you weren't getting any syntax errors in fn_loadPrices.sqf? I think the undefined variables in those expressions are because loadPrices wasn't executing properly

**Comment by Jawshy** [comment]
@dbenson24, I was also getting errors for fn_loadPrices.sqf; I just didn't notice it higher up in the log. 

I'll be committing a SpyGlass config update momentarily.

**Comment by dbenson24** [comment]
@Jawshy Just curious, did you do any testing yet?

**Comment by Jawshy** [comment]
Yes. That's how I got those above errors.

**Comment by dbenson24** [comment]
@Jawshy sorry should have been more clear, I meant after I commited the syntax fix

**Comment by Jawshy** [comment]
Oh, now I understand. No, I haven't yet tested https://github.com/AsYetUntitled/Framework/pull/47/commits/fda59b550e6489c02229376d8eb0f4e6d221349d. I'm going to test another issue shortly then when that's done I'll test dynamicmarket again.

**Comment by Jawshy** [comment]
@dbenson24 

The following was written to the server.rpt after End of Altis Life Server Init: 

```
2016/08/08, 16:03:59 Error in expression <;
};

_goods = [];

{
if ((_x select 1) != 0) then {
_name = format["%1MarketGoo>
2016/08/08, 16:03:59   Error position: <!= 0) then {
_name = format["%1MarketGoo>
2016/08/08, 16:03:59   Error Generic error in expression
2016/08/08, 16:03:59 File life_server\Functions\Economy\fn_syncPrices.sqf, line 26
```

Virtual items shops are empty. 

More variables to whitelist in SpyGlass.

**Comment by dbenson24** [comment]
@Jawshy thanks for the help with spyglass

@coresync2k the most recent commit I made should fix any issues that you were having

**Comment by DarkSilencerZA** [comment]
If the below is set to false, is the SQL stuff needed?

dynamic_market_persistence = false;

**Comment by dbenson24** [comment]
@DarkSilencerZA right now the non persistent setting still uses the database to load from at the server start. If I can figure out how to loop through all of the config options located in the vitem config it won't rely on that table anymore

**Comment by tkcjesse** [comment]
@dbenson24 I'll be working on this stuff tonight. To an extent.

**Comment by dbenson24** [comment]
... what is going on

**Comment by tkcjesse** [comment]
Um wtf. Something terrible is happening with the checks... lol

**Comment by CarlosHingel** [comment]
lol

**Comment by tkcjesse** [comment]
Oh my fuck. What the fuck.

**Comment by setoy** [comment]
Yeah, well, thanks Outlook. Nice mails there :D

**Comment by CarlosHingel** [comment]
Wtf is going on.

**Comment by tkcjesse** [comment]
May want to unsubscribe from this bad boy for a bit. Lol.

**Comment by setoy** [comment]
Well, I'ma go to bed now... Let's see if this thing is still scrollable when I wake up :D

**Comment by CarlosHingel** [comment]
Now that this stopped, i think someone should add the right tags back to this and re-open it.

**Comment by dbenson24** [comment]
I'm gonna delete this PR and swap to the squashed PR I just made on my real fork of this repo

**Comment by DarkSilencerZA** [comment]
That was rather interesting to read through :dancers:

**Pull request closed without merging**